src/Lean/Elab/Coinductive.lean

```diff
@@ -0,0 +1,467 @@
+  | Name.str p s => Name.str p (s ++ "_functor")
+  | Name.num p n => Name.num (addFunctorPostfix p) n
+
+public def removeFunctorPostfix : Name → Name
+  | Name.anonymous => Name.anonymous
+  | Name.str p s => Name.str p (s.stripSuffix "_functor")
+  | Name.num p n => Name.num (removeFunctorPostfix p) n
+
+public def removeFunctorPostfixInCtor : Name → Name
+  | Name.anonymous => Name.anonymous
+  | Name.str p s => Name.str (removeFunctorPostfix p) s
+  | Name.num p n => Name.num (removeFunctorPostfixInCtor p) n
+
+private def rewriteGoalUsingEq (goal : MVarId) (eq : Expr) (symm : Bool := false) : MetaM MVarId := do
+  let rewriteResult ← goal.rewrite (←goal.getType) eq symm
+  goal.replaceTargetEq rewriteResult.eNew rewriteResult.eqProof
+
+/--
+  Generates unrolling lemmas that relate coinductive predicates to their flat inductive forms.
+  These lemmas are essential for the constructor generation process, providing the bridge
+  between the user-facing coinductive predicates and their internal flat representations.
+-/
+private def generateEqLemmas (infos : Array InductiveVal) : MetaM Unit := do
+  let levels := infos[0]!.levelParams.map mkLevelParam
+  for info in infos do
+    let res ← forallTelescopeReducing info.type fun args _ => do
+      let params := args.take <| info.numParams - infos.size
+      let args := args.extract info.numParams
+
+      let lhs := mkConst (removeFunctorPostfix info.name) levels
+      let lhs := mkAppN lhs params
+      let lhs := mkAppN lhs args
+
+      let calls := infos.map fun info => mkAppN (mkConst (removeFunctorPostfix info.name) levels) params
+      let rhs := mkConst info.name levels
+      let rhs := mkAppN rhs (params ++ calls ++ args)
+
+      let goalType ← mkEq lhs rhs
+      let goal ← mkFreshExprMVar goalType
+
+      let goalMVarId := goal.mvarId!
+
+      let .some #[fixEq] ←  getEqnsFor? (removeFunctorPostfix info.name) | throwError "did not generate unfolding theorem"
+      let existentialEquiv := mkConst (info.name ++ `existential_equiv) levels
+
+      let mut fixEq := mkConst fixEq levels
+      fixEq := mkAppN fixEq params
+      for arg in args do
+        fixEq ← mkCongrFun fixEq arg
+
+      let newGoal ← rewriteGoalUsingEq goalMVarId existentialEquiv
+      newGoal.assign fixEq
+
+      let goal ← instantiateMVars goal
+      mkLambdaFVars (params ++ args) goal
+    trace[Elab.coinductive] "res: {res}"
+    addDecl <| .defnDecl (←mkDefinitionValInferringUnsafe
+      ((removeFunctorPostfix info.name) ++ `functor_unfold) info.levelParams (←inferType res) res .opaque)
+
+/--
+  Generates a constructor for a coinductive predicate that corresponds to constructors
+  in the original `InductiveView`.
+
+  The process:
+  1. Takes the flat inductive constructor type
+  2. Fills recursive call parameters with the actual coinductive predicates
+  3. Converts to existential form using the equivalence lemma
+  4. Applies the unrolling rule to get the final constructor form
+-/
```

Comment:
I’m a bit surprised that you are calculating the type here from the flat inductive constructor, instead of taking the type that the user specified originally. Maybe it doesn't matter, just seems a bit round-about.